File: dotnet-service/src/Internal/Domain/Entities/Product.cs
using DotnetService.Internal.Domain.Errors;

namespace DotnetService.Internal.Domain.Entities;

public sealed record Product
{
    public string Id { get; init; } = string.Empty;
    public string Name { get; init; } = string.Empty;
    public decimal Price { get; init; }
    public IReadOnlyCollection<string> Tags { get; init; } = Array.Empty<string>();
    public DateTime CreatedAt { get; init; }
    public DateTime UpdatedAt { get; init; }

    public Product Validate()
    {
        if (string.IsNullOrWhiteSpace(Name))
        {
            throw DomainException.Validation("name is required");
        }

        if (Price <= 0)
        {
            throw DomainException.Validation("price must be positive");
        }

        return this;
    }
}

File: node-service/internal/domain/product.js
const DomainError = require('./errors');

class Product {
  constructor({ id, name, price, tags = [], createdAt, updatedAt }) {
    this.id = id;
    this.name = name;
    this.price = price;
    this.tags = tags || [];
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
  }

  validate() {
    if (!this.name || !this.name.trim()) {
      throw DomainError.validation('name is required');
    }
    if (typeof this.price !== 'number' || this.price <= 0) {
      throw DomainError.validation('price must be a positive number');
    }
    if (!Array.isArray(this.tags)) {
      this.tags = [];
    }
  }
}

module.exports = Product;
